### DISCUSSION TITLE
How to load several timestamps at once?

### DISCUSSION THREAD

**Discussion by Cadene** [python-ffmpegio/python-ffmpegio#42] [Q&A]
How to do it directly with ffmpegio without a for loop?
```python
    list_frames = []
    for timestamp in timestamps:
        fs, frames = ffmpegio.video.read(
            str(video_path), ss=str(timestamp), pix_fmt=image_format
        )
        list_frames.append(frames)
```

Thanks 🙏

**Comment by tikuma-lsuhsc** [comment]
### Contiguous frames

You have a couple options:

1. Use `vframes` (or equivalently `frames:v`) to read `n` frames at once:

```python
fs, frames = ffmpegio.video.read(video_path, ss=timestamps[0], vframes=len(timestamps), pix_fmt=image_format)
```

2. Use `to` to specify the ending timestamp:

```python
fs, frames = ffmpegio.video.read(video_path, ss=timestamps[0], to=timestamps[-1], pix_fmt=image_format)
```

Caveat: I'm not 100% sure if this will always include the last frame at `to`, i.e., whether to is an inclusive or exclusive end point. I recommend using the first option.

### Non-contiguous frames

#### If the # of keeper frames >> # of skipped frames

You can just slice the `frames` to your liking

```python
fs, frames = ffmpegio.video.read(video_path, ss=timestamps[0], to=timestamps[-1], pix_fmt=image_format)

n = np.round(timestamps*fs).astype(int)
frames = frames[n-n[0], ...]
```

This could very well be the fastest way regardless of how much you're keeping if you're using CUDA as you implied on your other thread.

#### If the # of keeper frames << # of skipped frames

You can use [select filter](https://ffmpeg.org/ffmpeg-filters.html#select_002c-aselect) to specify which frames to return. Learn [the FFmpeg expression](https://ffmpeg.org/ffmpeg-utils.html#Expression-Evaluation) from its documentation to formulate your selection logic, but using one of the examples:

```python
vf_expr = ''select='not(mod(n\,100))'" # every 100 frames
fs, frames = ffmpegio.video.read(video_path, ss=timestamps[0], vframes=len(timestamps), pix_fmt=image_format, vf=vf_expr)
```

Use the `vframes` option in conjunction so FFmpeg stops immediately after retrieving the last frame.

If your timestamps are randomly chosen (as in no formula can describe your choice) then your expression could get too long. Also, there are several timestamp related options that you may need to set for it to return timestamp accurate frames (e.g., `copyts`). Nonetheless, give it a try first.

**Reply by tikuma-lsuhsc** [reply]
**[This solution does not work currently as v0.9 does not correctly escape commas. Wait till v0.10 release for the fix]**

One more thing. FFmpeg filtergraph expression must be properly escaped. If you want to avoid dealing with it, you can try using `ffmpegio.filtergraph` submodule. For the example filter above, you can define it by:

```python
vf_expr = ffmpegio.filtergraph.select('not(mod(n,100))')
```

Note it takes `'not(mod(n,100))'` instead of `'not(mod(n\,100))'`

**Reply by Cadene** [reply]
What an incredibly detailed answer. Thank you so much. I will try this and come back to you ;)

**Reply by tikuma-lsuhsc** [reply]
You're welcome. Remember that the reader/writer functions take essentially any FFmpeg command option as a keyword argument.
